### PR TITLE
reformat iv selection class

### PR DESCRIPTION
**Pull request by solegalli**
Hi @Morgan-Sell 

I reformatted the IV selector quite a bit, capturing the abstractions in a separate class.

Please have a look and if you like it, we would need to adapt the tests.

Also, as usual, I rebased main, therefore all the file changes.

Thank you!

### CODE REVIEW

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

Ok, wow, you're superwoman! You've done a done of work. 

It looks good to me! I'm a bit confused as to why there are conflicts.

I'm a bit slow today. The jet lag hit me hard this morning.

**Comment by solegalli** [comment]
Resolved. 

You will probably have more conflicts because I made 3 or 4 major code refactoring PR merges. Have a look and see if you can troubleshoot them :)

File: feature_engine/encoding/woe.py [hunk 1]
```diff
@@ -30,6 +30,44 @@
...
+        if y.min() != 0 or y.max() != 1:
+            y = pd.Series(np.where(y == y.min(), 0, 1))
+        return X, y
+
+    def _calculate_woe(self, X: pd.DataFrame, y: pd.Series, variable: str):
+        total_pos = y.sum()
+        inverse_y = y.ne(1).copy()
+        total_neg = inverse_y.sum()
+
+        pos = y.groupby(X[variable]).sum() / total_pos
```
Comment (Morgan-Sell) [inline feature_engine/encoding/woe.py:58]: How does this work? `y` is a panda series. Doesn't groupby applied to dataframes? Also, should the `by` parameter be a string that is the name of a variable?

Should we concatenate `X`, `y`, and `inverse_y` to create `main`? Imagine `y` and `inverse_y` are variables in the `main` dataframe. If so, the formula would be:

`pos = main.groupby(variable).sum()["y"] / total_pos`

File: feature_engine/encoding/woe.py [hunk 2]
```diff
@@ -30,6 +30,44 @@
...
+        neg = inverse_y.groupby(X[variable]).sum() / total_neg
+
```
Comment (Morgan-Sell) [inline feature_engine/encoding/woe.py:60]: The same question about the groupby functionality used to calculate `pos`, but using `inverse_y`.

Proposed formula:

`neg = main.groupby(variable).sum()["inverse_y"] / total_neg`

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

I prematurely merged. I was working on building the unit test and came across the above issues. What do you think?

File: feature_engine/encoding/woe.py (hunk 1 above)
Comment (solegalli) [reply feature_engine/encoding/woe.py:58]: they work fine. They should not be changed. 

Did you try it out in a jupyter notebook?

**Comment by solegalli** [comment]
I don't think it matters if you merge now or later.

They key thing is to have the branch synced with the one from where you are making the PR to the main repo, and keep adding the changes there.

The formulas for pos and neg are working as expected, because we have tests for the woe in place and they return the expected values. So I would not change those.

I think I also added all the logic. We just need the tests and the documentation.

Maybe some of the tests for the woe may help for the tests for this class?

File: feature_engine/encoding/woe.py (hunk 1 above)
Comment (Morgan-Sell) [reply feature_engine/encoding/woe.py:58]: Perdoname! Soy un boludo!

I did test it in a notebook. The issue was that the dataset had a dependent variable comprised of yes and no, which I mapped to 0 and 1. However, I didn't call the revised/mapped "y" feature when I was testing the WoE class. Ugh!

**Pull request merged**